hexf/main.swift

```diff
@@ -0,0 +1,188 @@
+        
+        do {
+            // TODO: Heed deprecation warning, and get right config type. This will require an availability check.
+            //            try NSWorkspace.shared.openApplication(at: url, configuration: config)
+            try NSWorkspace.shared.launchApplication(at: url, options: NSWorkspace.LaunchOptions.default, configuration: config)
+        } catch {
+            fputs("Launch app failed: \(error.localizedDescription)", stderr)
+            throw HexfError.launchAppFailure
+        }
+    }
+    
+    private func processStandardInput() throws {
+        let inFile = FileHandle.standardInput
+        // TODO: Heed deprecation warning. This will require an availability check.
+        let data = inFile.readDataToEndOfFile()
+        
+        guard data.count != 0 else {
+            throw HexfError.standardInputNoData
+        }
+        
+        if self.appRunning {
+            // App is already running so post distributed notification
+            let center = DistributedNotificationCenter.default()
+            center.postNotificationName(NSNotification.Name(rawValue: "HFOpenDataNotification"),
+                                        object: nil,
+                                        userInfo: ["data" : data],
+                                        deliverImmediately: true)
+            return
+        }
+        
+        // App isn't running so launch it with custom args
+        try launchApp(with: [
+            "-HFOpenData",
+            data.base64EncodedString(options: .init(rawValue: 0))
+            // I'm not sure what 0 is for the options, but that is what the Obj-C code passed in... so I kept it.
+            //        NSString *base64Str = [data base64EncodedStringWithOptions:0];
+        ])
+    }
+    
+    enum Commands: Equatable {
+        case diff(leftFile: String, rightFile: String)
+        case open(files: [String])
+    }
+    
+    enum Options: Equatable {
+        case command(_ command: Commands)
+        case none
+        case help
+        case invalid
+        
+        init(args: [String]) {
+            if args.count == 1 {
+                self = .none
+            } else if args.count == 4, args[1] == "-d" {
+                self = .command(.diff(leftFile: standardize(path:args[2]), rightFile: standardize(path:args[3])))
+            } else {
+                var filesToOpen = [String]()
+                for arg in args.dropFirst() {
+                    if arg.hasPrefix("-") {
+                        if arg == "-h" || arg == "--help" {
+                            self = .help
+                            return
+                        }
+                        self = .invalid
+                        return
+                    }
+                    filesToOpen.append(standardize(path: arg))
+                }
+                self = .command(.open(files: filesToOpen))
+            }
+        }
+    }
+    
+    func process(arguments args: [String]) throws {
```

Comment:
what do you think about using apple's library for this?
https://github.com/apple/swift-argument-parser

Reply:
I did consider that. I think at the time I decided against it because I wasn't sure what the project's policy was on dependencies.
I'll bring it in!